cpp/src/binaryop/compiled/binary_ops.cu

```diff
@@ -239,6 +245,93 @@ struct null_considering_binop {
+    ? thrust::tabulate(rmm::exec_policy(stream),
+                       out.begin<bool>(),
+                       out.end<bool>(),
+                       [optional_iter, compare] __device__(size_type i) {
+                         return optional_iter[i].has_value() and compare(i, i);
+                       })
+    : thrust::tabulate(rmm::exec_policy(stream),
+                       out.begin<bool>(),
+                       out.end<bool>(),
+                       [optional_iter, compare] __device__(size_type i) {
+                         return optional_iter[i].has_value() and not compare(i, i);
+                       });
+}
+
+void struct_binary_operation(mutable_column_view& out,
+                             column_view const& lhs,
+                             column_view const& rhs,
+                             binary_operator op,
+                             rmm::cuda_stream_view stream)
+{
+  bool const has_struct_nulls =
+    structs::detail::contains_struct_nulls(lhs) || structs::detail::contains_struct_nulls(rhs);
+  auto const lhs_flattened = structs::detail::flatten_nested_columns(
+    table_view{{lhs}},
+    {},
+    {},
+    has_struct_nulls ? structs::detail::column_nullability::FORCE
+                     : structs::detail::column_nullability::MATCH_INCOMING);
+  auto const rhs_flattened = structs::detail::flatten_nested_columns(
+    table_view{{rhs}},
+    {},
+    {},
+    has_struct_nulls ? structs::detail::column_nullability::FORCE
+                     : structs::detail::column_nullability::MATCH_INCOMING);
+
+  auto d_lhs = table_device_view::create(lhs_flattened);
+  auto d_rhs = table_device_view::create(rhs_flattened);
+  auto d_out = column_device_view::create(out, stream);
+  auto out_iter =
+    cudf::detail::make_optional_iterator<bool>(*d_out, contains_nulls::DYNAMIC{}, out.has_nulls());
+  bool has_nulls = has_nested_nulls(lhs_flattened) || has_nested_nulls(rhs_flattened);
+
+  if (op == binary_operator::EQUAL || op == binary_operator::NOT_EQUAL) {
+    has_nulls ? struct_compare_tabulation(
+                  out, row_equality_comparator<true>{*d_lhs, *d_rhs, true}, op, out_iter, stream)
+              : struct_compare_tabulation(
+                  out, row_equality_comparator<false>{*d_lhs, *d_rhs, true}, op, out_iter, stream);
+  } else if (op == binary_operator::LESS || op == binary_operator::LESS_EQUAL ||
+             op == binary_operator::GREATER || op == binary_operator::GREATER_EQUAL) {
+    auto const num_columns = lhs_flattened.flattened_columns().num_columns();
+    has_nulls
+      ? struct_compare_tabulation(out,
+                                  row_lexicographic_comparator<true>{
+                                    *d_lhs, *d_rhs, get_orders(op, num_columns, stream).data()},
+                                  op,
+                                  out_iter,
+                                  stream)
```

Comment:
This can be extracted to a (template) lambda putting on top of these lines. Something like:
```
auto do_tabulation = [&] <bool has_null> (auto const& ) {
struct_compare_tabulation(...);
};
```

Then you just call one simple line multiple times (with different parameter).

Reply:
I actually can't get this templated lambda function to compile and can't find another case where this templating is used in cudf. Have I missed something or does this not work?

Reply:
Explicit template lambdas are a C++20 feature. As of C++14 you can write generic lambdas using auto deduction of parameter types and return types, but without C++20 explicit template params like this won't work and libcudf is only compiled with C++17 for now.